src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,713 @@
+    fprintf(f, "  -I%s \\\n", java_include.c_str());
+    fprintf(f, "  -I%s/%s \\\n", java_include.c_str(), config.java_md_.c_str());
+    fprintf(f, "  -I%s \\\n", include_.c_str());
+    bool have_oflag = (strstr(options_.c_str(), "-O") != nullptr);
+    fprintf(f, "  %s%s \\\n", have_oflag ? "" : "-O3", options_.c_str());
+    fprintf(f, "  -DHAIL_MODULE=$(MODULE)\n");
+    fprintf(f, "LIBFLAGS := -fvisibility=default %s\n", 
+      config.is_darwin_ ? "-dynamiclib -Wl,-undefined,dynamic_lookup"
+                         : "-rdynamic -shared");
+    fprintf(f, "\n");
+    // top target is the .so
+    fprintf(f, "$(MODULE_SO): $(MODULE).o\n");
+    fprintf(f, "\t-[ -f $(MODULE).lockX ] || /usr/bin/touch $(MODULE).lockX ; \\\n");
+    fprintf(f, "\t  while ! /bin/ln $(MODULE).lockX $(MODULE).lock 2>/dev/null ; do sleep 0.1 ; done ; \\\n");
+    fprintf(f, "\t  /bin/ln -f $(MODULE).new $@ ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).new ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).lock\n");
+    fprintf(f, "\n");
+    // build .o from .cpp
+    fprintf(f, "$(MODULE).o: $(MODULE).cpp\n");
+    fprintf(f, "\t$(CXX) $(CXXFLAGS) -o $@ -c $< 2> $(MODULE).err && \\\n");
+    fprintf(f, "\t  $(CXX) $(CXXFLAGS) $(LIBFLAGS) -o $(MODULE).tmp $(MODULE).o 2>> $(MODULE).err ; \\\n");
+    fprintf(f, "\tstatus=$$? ; \\\n");
+    fprintf(f, "\tif [ $$status -ne 0 ] || [ -z $(MODULE).tmp ]; then \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).new ; \\\n");
+    fprintf(f, "\t  echo FAIL ; exit 1 ; \\\n");
+    fprintf(f, "\tfi\n");
+    fprintf(f, "\t-[ -f $(MODULE).lockX ] || /usr/bin/touch $(MODULE).lockX ; \\\n");
+    fprintf(f, "\t  while ! /bin/ln $(MODULE).lockX $(MODULE).lock 2>/dev/null ; do sleep 0.1 ; done ; \\\n");
+    fprintf(f, "\t  /bin/ln -f $(MODULE).tmp $(MODULE).new ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).tmp ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).err ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).lock\n");
+    fprintf(f, "\n");
+    fclose(f);
+  }
+
+public:
+  bool try_to_start_build() {
+    // Try to create the .new file
+    std::lock_guard<NativeModule> mylock(*parent_);
+    int fd = ::open(hm_new_.c_str(), O_WRONLY|O_CREAT|O_EXCL, 0666);
+    if (fd < 0) {
+      // We lost the race to start the build
+      return false;
+    }
+    ::close(fd);
+    // The .new file may look the same age as the .cpp file, but
+    // the makefile is written to ignore the .new timestamp
+    write_mak();
+    write_cpp();
+    std::stringstream ss;
+    ss << "/usr/bin/make -B -C " << config.module_dir_ << " -f " << hm_mak_;
+    ss << " 1>/dev/null &";
+    int rc = system(ss.str().c_str());
+    if (rc < 0) perror("system");
+    return true;
+  }
+};
+
+NativeModule::NativeModule(
+  const char* options,
+  const char* source,
+  const char* include,
+  bool force_build
+) :
+  build_state_(kInit),
+  load_state_(kInit),
+  key_(hash_two_strings(options, source)),
+  is_global_(false),
```

Comment:
what is `is_global_`?

Reply:
I wanted to have a way to access functions which were already in libhail.so and call them
through the same Scala NativeLongFuncN/NativePtrFuncN interfaces - at least for testing.
Rather than replicating the whole function-query interface, there's a special "global" module which can be used for accessing functions from non-dynamically-compiled code (and without
applying the same name-mangling which keeps modules in their own namespaces).